glib/src/object.rs

```diff
@@ -3267,6 +3294,87 @@ wrapper! {
     }
 }
 
+struct WeakRefNotifyData {
+    callback: Box<dyn Fn()>,
+    should_drop: AtomicBool,
+}
+
+impl WeakRefNotifyData {
+    pub fn new<F: Fn() + 'static>(f: F) -> ManuallyDrop<Pin<Box<Self>>> {
+        ManuallyDrop::new(Box::pin(WeakRefNotifyData {
+            callback: Box::new(f),
+            should_drop: AtomicBool::new(false),
+        }))
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Checks whether the data should be dropped and marks it to be dropped
+    pub fn should_drop(&self) -> bool {
+        self.should_drop.swap(true, Ordering::SeqCst)
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// A handle to disconnect a weak ref notify closure.
+pub struct WeakRefNotify {
+    object: WeakRef<Object>,
+    data: ManuallyDrop<Pin<Box<WeakRefNotifyData>>>,
+    func: gobject_ffi::GWeakNotify,
+}
+
+impl WeakRefNotify {
+    pub(crate) fn new<T: ObjectType, F: Fn() + 'static>(obj: &T, f: F) -> WeakRefNotify {
+        unsafe extern "C" fn notify_func(data: ffi::gpointer, _obj: *mut gobject_ffi::GObject) {
+            let data = Box::from_raw(data as *mut WeakRefNotifyData);
+
+            (data.callback)();
+
+            // Don't drop the data when the `WeakRefNotify` handle wasn't dropped yet
+            if !data.should_drop() {
+                Box::leak(data);
+            }
+        }
+
+        unsafe {
+            let mut data = WeakRefNotifyData::new(f);
+            let obj_ptr = obj.as_object_ref().to_glib_none().0;
+
+            let data_ptr: *mut _ = Pin::as_mut(&mut data).get_unchecked_mut();
+            gobject_ffi::g_object_weak_ref(obj_ptr, Some(notify_func), data_ptr as *mut _);
+
+            WeakRefNotify {
+                object: obj.unsafe_cast_ref::<Object>().downgrade(),
+                data,
+                func: Some(notify_func),
+            }
+        }
+    }
+
+    #[doc(alias = "g_object_weak_unref")]
+    pub fn disconnect(self) {
+        unsafe {
+            if !self.data.should_drop() {
+                if let Some(obj) = self.object.upgrade() {
```

Comment:
I think this should be before the `should_drop()` or not? Otherwise you might end up swapping the bool but then upgrading fails and then you leak the data?